### PR TITLE
[RLlib] Introduce new policy base classes.

### PR DESCRIPTION
**Pull request by gjoliver**
With base functionality and API definitions (borrowed faithfully from build_xx_policy()).

## Why are these changes needed?

We are updating RLlib's policies to use sub-classing design.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [*] Unit tests
   - [*] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by gjoliver** [comment]
This should be safe to merge.
We introduce these base classes that are not being used by anyone first.

**Review by sven1977 — APPROVED** [review]
Looks great. Let's merge! Already reviewed this in the larger PR (that resulted in splitting).
Thanks @gjoliver .

**Review by smorad — COMMENTED** [review]
I left a ton of nitpicks on the name changes, please don't be annoyed. As a user, the complexity of the old `TorchPolicy` class scared me from ever overriding any functions or implementing my own custom policies.

If you are rewriting the policy classes, I think it's very important to make these classes more readable than the original ones. Like knowing which of the compute actions functions call each other. IMO, there are 3 main operations the policy is responsible for: it computes logits/actions in inference mode during rollouts, it computes a loss on train batches, and it optimizes the model using the loss. The number of functions here suggests the policy class is doing way more than this.

This is all really philosophical stuff, detached from implementation and the real-world. But maybe you could let me know next time you and @sven1977 discuss this and I could listen in?

File: rllib/policy/torch_policy_v2.py [hunk 1]
```diff
@@ -0,0 +1,1220 @@
...
+            train_batch: The training data.
+
+        Returns:
+            Loss tensor given the input batch.
+        """
+        raise NotImplementedError
+
+    @DeveloperAPI
+    @OverrideToImplementCustomLogic
+    def action_sampler_fn(
```
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/policy/torch_policy_v2.py:247]: +1

File: rllib/policy/torch_policy_v2.py [hunk 2]
```diff
@@ -0,0 +1,1220 @@
...
+            Sampled action
+            Log-likelihood
+            Action distribution inputs
+            Updated state
+        """
+        return None, None, None, None
+
+    @DeveloperAPI
+    @OverrideToImplementCustomLogic
+    def action_distribution_fn(
```
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/policy/torch_policy_v2.py:272]: +1

File: rllib/policy/torch_policy_v2.py [hunk 3]
```diff
@@ -0,0 +1,1220 @@
...
+
+        Returns:
+            Size N. A sample batch must be of size K*N.
+        """
+        # By default, any sized batch is ok, so simply return 1.
+        return 1
+
+    @DeveloperAPI
+    @OverrideToImplementCustomLogic
+    def stats_fn(self, train_batch: SampleBatch) -> Dict[str, TensorType]:
```
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/policy/torch_policy_v2.py:333]: +1

File: rllib/policy/torch_policy_v2.py [hunk 4]
```diff
@@ -0,0 +1,1220 @@
...
+            train_batch: The SampleBatch (already) used for training.
+
+        Returns:
+            The stats dict.
+        """
+        return {}
+
+    @DeveloperAPI
+    @OverrideToImplementCustomLogic_CallToSuperRecommended
+    def extra_grad_process(
```
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/policy/torch_policy_v2.py:346]: +1

File: rllib/policy/torch_policy_v2.py [hunk 5]
```diff
@@ -0,0 +1,1220 @@
...
+        # Create n sample-batch buffers (num_multi_gpu_tower_stacks), each
+        # one with m towers (num_gpus).
+        num_buffers = self.config.get("num_multi_gpu_tower_stacks", 1)
+        self._loaded_batches = [[] for _ in range(num_buffers)]
+
+        # If set, means we are using distributed allreduce during learning.
+        self.distributed_world_size = None
+
+        self.batch_divisibility_req = self.get_batch_divisibility_req()
+        self.max_seq_len = max_seq_len
```
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/policy/torch_policy_v2.py:222]: I'm confused as to why we need to explicitly pass `max_seq_len` to the policy. Can't we just pass it through the config object?

File: rllib/policy/torch_policy_v2.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/policy/torch_policy_v2.py:247]: It also seems like the signature is copied from the previous ugly functional-API. But with this new base class we can actually now leverage `self.model`. So there is no need to pass in `ModelV2` objects anymore?

File: rllib/policy/torch_policy_v2.py [hunk 6]
```diff
@@ -0,0 +1,1220 @@
...
+            loss: The loss tensor associated with the optimizer.
+
+        Returns:
+            An dict with information on the gradient processing step.
+        """
+        return {}
+
+    @DeveloperAPI
+    @OverrideToImplementCustomLogic_CallToSuperRecommended
+    def extra_compute_grad_fetches(self) -> Dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/policy/torch_policy_v2.py:366]: +1

File: rllib/policy/torch_policy_v2.py [hunk 7]
```diff
@@ -0,0 +1,1220 @@
...
+
+        Returns:
+            Extra fetch dict to be added to the fetch dict of the
+            `compute_gradients` call.
+        """
+        return {LEARNER_STATS_KEY: {}}  # e.g, stats, td error, etc.
+
+    @DeveloperAPI
+    @OverrideToImplementCustomLogic_CallToSuperRecommended
+    def extra_action_out(
```
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/policy/torch_policy_v2.py:377]: +1

File: rllib/policy/torch_policy_v2.py [hunk 8]
```diff
@@ -0,0 +1,1220 @@
...
+            ]
+        else:
+            optimizers = [torch.optim.Adam(self.model.parameters())]
+        if getattr(self, "exploration", None):
+            optimizers = self.exploration.get_exploration_optimizer(optimizers)
+        return optimizers
+
+    def _init_model_and_dist_class(self):
+        if is_overridden(self.make_model) and is_overridden(
+            self.make_model_and_action_dist
```
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/policy/torch_policy_v2.py:447]: I think `is_overridden()` is missing from the PR? I can't find the function

File: rllib/policy/torch_policy_v2.py [hunk 9]
```diff
@@ -0,0 +1,1220 @@
...
+                obs_space=self.observation_space,
+                action_space=self.action_space,
+                num_outputs=logit_dim,
+                model_config=self.config["model"],
+                framework=self.framework,
+            )
+        return model, dist_class
+
+    @override(Policy)
+    def compute_actions_from_input_dict(
```
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/policy/torch_policy_v2.py:475]: Do we have a design doc for this? It's also very confusing for me. I can see we have different modes of calling the policy during exploration, training, inference, and even different modes during training (e.g. SAC). We also need to do post/preprocessing of data before this stuff. We should take a step back and simplify this design.

File: rllib/policy/torch_policy_v2.py [hunk 10]
```diff
@@ -0,0 +1,1220 @@
...
+        return len(self.model.get_initial_state())
+
+    @override(Policy)
+    @DeveloperAPI
+    def get_initial_state(self) -> List[TensorType]:
+        return [s.detach().cpu().numpy() for s in self.model.get_initial_state()]
+
+    @override(Policy)
+    @DeveloperAPI
+    def get_state(self) -> Union[Dict[str, TensorType], List[TensorType]]:
```
[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/policy/torch_policy_v2.py:896]: +1

[the remark this answers is not in the document]
Comment (kouroshHakha) [reply rllib/policy/torch_policy_v2.py:896]: or `get_state_dict` that has both model states and optimizer state?

File: rllib/policy/torch_policy_v2.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/policy/torch_policy_v2.py:222]: I believe this is for differentiating multiple sequences from different episodes in a single batch.
so we can do all the post-processing (currently lives on Policy) correctly.
I think max_seq_len is a great topic for deep dive actually. I am also confused about it.

File: rllib/policy/torch_policy_v2.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/policy/torch_policy_v2.py:247]: yeah, good point. let me do a round of cleanups based on the comments here.

File: rllib/policy/torch_policy_v2.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/policy/torch_policy_v2.py:272]: so I want to remind us there is also value to have backward compatible names, so users don't feel like everything is changing all at once, and they don't know what to do.
also, having a consistent scheme where things you can override always have _fn postfix feels nice.

File: rllib/policy/torch_policy_v2.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/policy/torch_policy_v2.py:377]: a lot of these are because of the TF1 policies.
you only have to pre-specify outs and fetches when you are dealing with static graph, for example.

File: rllib/policy/torch_policy_v2.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/policy/torch_policy_v2.py:447]: good point man. huh, let me merge those utils too.

File: rllib/policy/torch_policy_v2.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/policy/torch_policy_v2.py:475]: yeah, I think Sven wants to only keep compute_actions_from_input_dict.
can't wait to clean this up. but probably can only happen after we migrate all the policies?

File: rllib/policy/torch_policy_v2.py [hunk 11]
```diff
@@ -0,0 +1,1220 @@
...
+            return log_likelihoods
+
+    @with_lock
+    @override(Policy)
+    @DeveloperAPI
+    def learn_on_batch(self, postprocessed_batch: SampleBatch) -> Dict[str, TensorType]:
+
+        # Set Model to train mode.
+        if self.model:
+            self.model.train()
```
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/policy/torch_policy_v2.py:616]: not sure man. this is not new code. need to study.

File: rllib/policy/torch_policy_v2.py [hunk 12]
```diff
@@ -0,0 +1,1220 @@
...
+    @override(Policy)
+    @DeveloperAPI
+    def get_num_samples_loaded_into_buffer(self, buffer_index: int = 0) -> int:
+        if len(self.devices) == 1 and self.devices[0] == "/cpu:0":
+            assert buffer_index == 0
+        return len(self._loaded_batches[buffer_index])
+
+    @override(Policy)
+    @DeveloperAPI
+    def learn_on_loaded_batch(self, offset: int = 0, buffer_index: int = 0):
```
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/policy/torch_policy_v2.py:701]: probably. what about CPU case?

File: rllib/policy/torch_policy_v2.py [hunk 13]
```diff
@@ -0,0 +1,1220 @@
...
+
+        # Do the (maybe parallelized) gradient calculation step.
+        tower_outputs = self._multi_gpu_parallel_grad_calc(device_batches)
+
+        # Mean-reduce gradients over GPU-towers (do this on CPU: self.device).
+        all_grads = []
+        for i in range(len(tower_outputs[0][0])):
+            if tower_outputs[0][0][i] is not None:
+                all_grads.append(
+                    torch.mean(
```
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/policy/torch_policy_v2.py:762]: hmm, wait, torch.mean([g1, g2, g3]) is the average gradient right?
the scale of the gradient should be the same? it's like manual all-reduce.
maybe I am missing something.

File: rllib/policy/torch_policy_v2.py [hunk 14]
```diff
@@ -0,0 +1,1220 @@
...
+
+        postprocessed_batch.set_training(True)
+        self._lazy_tensor_dict(postprocessed_batch, device=self.devices[0])
+
+        # Do the (maybe parallelized) gradient calculation step.
+        tower_outputs = self._multi_gpu_parallel_grad_calc([postprocessed_batch])
+
+        all_grads, grad_info = tower_outputs[0]
+
+        grad_info["allreduce_latency"] /= len(self._optimizers)
```
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/policy/torch_policy_v2.py:812]: yeah, I think that's reasonable.

File: rllib/policy/torch_policy_v2.py [hunk 15]
```diff
@@ -0,0 +1,1220 @@
...
+                + ["state_ins", SampleBatch.SEQ_LENS],
+                output_names=["output", "state_outs"],
+                dynamic_axes={
+                    k: {0: "batch_size"}
+                    for k in list(dummy_inputs.keys())
+                    + ["state_ins", SampleBatch.SEQ_LENS]
+                },
+            )
+        else:
+            traced = torch.jit.trace(self.model, (dummy_inputs, state_ins, seq_lens))
```
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/policy/torch_policy_v2.py:976]: I agree. how can we be sure though? do you want to make a test, etc?

[the remark this answers is not in the document]
Comment (smorad) [reply rllib/policy/torch_policy_v2.py:976]: The current implementation is already incorrect, so a new attempt can't be worse.
Here is a simple test showing how to verify loading/storing arbitrary torch models:

```python
import torch
import copy

net = torch.nn.Sequential(
  torch.nn.Linear(8, 8),
  torch.nn.Linear(8, 8),
)
torch.save(copy.deepcopy(net), '/tmp/unittest_module')
net_load = torch.load('/tmpp/unittest_module')
for param_i, param_j in zip(net.parameters(), net_load.parameters()):
  assert torch.all(param_i == param_j)
```

I am willing to write a better one in master.

File: rllib/policy/torch_policy_v2.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (smorad) [reply rllib/policy/torch_policy_v2.py:701]: It should be the same as GPU case I think. The towers can be on GPU or CPU. 
```python
loss.backward()
optimizer.step()
```
is agnostic to the device the model is on.

File: rllib/policy/torch_policy_v2.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (smorad) [reply rllib/policy/torch_policy_v2.py:377]: Ah, I know very little about how tf works. Thanks.

File: rllib/policy/torch_policy_v2.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (smorad) [reply rllib/policy/torch_policy_v2.py:762]: Let's say we have 3 state/action samples that produce losses `0.1, 0.2, 0.3`. If we were to train all three samples using a single tower, the loss would be `loss = 0.1 + 0.2 + 0.3 = 0.6`. However, now let's say we compute the same exact losses using three towers. If we take the mean rather than sum: `loss = (0.1 + 0.2 + 0.3) / 0.3 = 0.2`.

So if we take the mean, then increasing the number of towers will actually "slow" down the learning. So if you change the number of towers, you would need to multiply the `lr` by `num_towers` to retain the same timestep efficiency.

[the remark this answers is not in the document]
Comment (XuehaiPan) [reply rllib/policy/torch_policy_v2.py:762]: > Let's say we have a rollout of three timesteps that produces losses 0.1, 0.2, 0.3. If we were to train the entire rollout using a single tower, the loss would be loss = 0.1 + 0.2 + 0.3 = 0.6.

The mostly used reduction method for training losses is `mean`. Each tower loss is averaged over the number of samples (`batch_size * seq_lens`). If we compute the loss using a single tower, the loss is also `loss = (0.1 + 0.2 + 0.3) / 3 = 0.2`.

On each tower, we reduce losses over the number of samples:

https://github.com/ray-project/ray/blob/de69b0d6d625615a58ac5a80d25f0cf81882b0f4/rllib/agents/ppo/ppo_torch_policy.py#L107-L110

https://github.com/ray-project/ray/blob/de69b0d6d625615a58ac5a80d25f0cf81882b0f4/rllib/agents/ppo/ppo_torch_policy.py#L136-L143

File: rllib/policy/torch_policy_v2.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/policy/torch_policy_v2.py:272]: I'm with all of you :D 

We SHOULD change these names (@kouroshHakha , @smorad ), but not in this PR (@gjoliver ).

Let's have an internal meeting/discussion on better:
* Trainer API
* Policy/Model API

Yes, `compute_actions_from_input_dict` is a bad name for such a top-level API method.

File: rllib/policy/torch_policy_v2.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/policy/torch_policy_v2.py:247]: See my comment below. Yes, we SHOULD change names, but not in this PR and only after thourough discussion with the entire team. This is something you only want to do once to not further confuse users. Also, all old names must remain supported (with warning) for some time!

File: rllib/policy/torch_policy_v2.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (smorad) [reply rllib/policy/torch_policy_v2.py:762]: In that case, then by taking the mean over towers we are taking a mean of means, which is different than the mean over the number of samples.

[the remark this answers is not in the document]
Comment (XuehaiPan) [reply rllib/policy/torch_policy_v2.py:762]: > In that case, then by taking the mean over towers we are taking a mean of means, which can be different than the mean over the number of samples if the number of samples varies between towers.

If the batch is evenly divided for all towers, the result of "mean" vs. "mean of mean" are identical. The condition is always true for non-RNN cases, but is not always fulfilled when we have `seq_lens`  for RNNs.

[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/policy/torch_policy_v2.py:762]: Hey @smorad , @XuehaiPan , and @gjoliver and everyone else reading this.
I completely agree with Steven's "scaredness" about our Policy classes. The amount and complexity of the different methods that one could/should/must-not override is insane. I'm prepping a WIP PR here that addresses this problem:
https://github.com/ray-project/ray/pull/25293

This will have to be iterated over and is not the final solution.
